modules/network/src/main/java/org/apache/ignite/internal/network/DefaultMessagingService.java

```diff
@@ -337,82 +337,167 @@ private List<ClassDescriptorMessage> prepareMarshal(NetworkMessage msg) throws E
             }
         });
     }
 
-    private void handleIncomingMessage(InNetworkObject obj) {
-        if (senderIdIsStale(obj)) {
-            LOG.info("Sender ID {} ({}) is stale, so skipping message handling: {}", obj.launchId(), obj.consistentId(), obj.message());
-            return;
-        }
+    private static void logMessageSkipDueToSenderLeft(InNetworkObject inNetworkObject) {
+        LOG.info("Sender ID {} ({}) is stale, so skipping message handling: {}",
+                inNetworkObject.launchId(), inNetworkObject.consistentId(), inNetworkObject.message()
+        );
+    }
 
-        NetworkMessage msg = obj.message();
-        DescriptorRegistry registry = obj.registry();
+    private boolean senderIdIsStale(InNetworkObject obj) {
+        return staleIdDetector.isIdStale(obj.launchId());
+    }
+
+    private void handleInvokeResponse(InNetworkObject inNetworkObject) {
+        unmarshalMessage(inNetworkObject);
+
+        InvokeResponse response = (InvokeResponse) inNetworkObject.message();
+        onInvokeResponse(response.message(), response.correlationId());
+    }
+
+    private void unmarshalMessage(InNetworkObject obj) {
         try {
-            msg.unmarshal(marshaller, registry);
+            obj.message().unmarshal(marshaller, obj.registry());
         } catch (Exception e) {
             throw new IgniteException("Failed to unmarshal message: " + e.getMessage(), e);
         }
-        if (msg instanceof InvokeResponse) {
-            InvokeResponse response = (InvokeResponse) msg;
-            onInvokeResponse(response.message(), response.correlationId());
-            return;
+    }
+
+    private Executor chooseExecutorFor(NetworkMessage payload, InNetworkObject obj, ExecutorChooser<NetworkMessage> chooser) {
+        if (wantsInboundPool(chooser)) {
+            return chooseExecutorInInboundPool(obj);
+        } else {
+            return chooser.choose(payload);
         }
+    }
 
-        Long correlationId = null;
-        NetworkMessage message = msg;
+    private Executor chooseExecutorInInboundPool(InNetworkObject obj) {
+        return inboundExecutors.stripeFor(obj.connectionIndex());
+    }
 
-        if (msg instanceof InvokeRequest) {
-            // Unwrap invocation request
-            InvokeRequest messageWithCorrelation = (InvokeRequest) msg;
-            correlationId = messageWithCorrelation.correlationId();
-            message = messageWithCorrelation.message();
+    /**
+     * Finishes unmarshalling the message and handles it on current thread on first handler. Also handles it with other
+     * handlers (second and so on) on executors chosen by their choosers.
+     */
+    private void handleStartingWithFirstHandler(
+            NetworkMessage payload,
+            @Nullable Long correlationId,
+            InNetworkObject obj,
+            HandlerContext firstHandlerContext,
+            Iterator<HandlerContext> remainingContexts
+    ) {
+        if (senderIdIsStale(obj)) {
```

Comment:
These 2 checks are made in different threads. First one is to skip handling as early as possible. But after that a task is put to an executor, and who knows how much time it will sit there awaiting for execution. During this time it might happen that the sender has already become stale, hence second check.

Do you think second check is redundant?